clusterloader2: import k8s.io/component-base/metrics/testutil.Histogram to compute quantile

ingvagabund opened the pull request:
On top of https://github.com/kubernetes/perf-tests/pull/978

- importing `k8s.io/component-base/metrics/testutil.Histogram` to compute quantile
- adding unit test to validate the quantile of `k8s.io/perf-tests/clusterloader2/pkg/measurement/util.Histogram` is computed correctly

ingvagabund commented:
/hold cancel

mm4tt commented:
/lgtm

Looks good, will check the cl2 test results before approving

ingvagabund commented:
/hold

ingvagabund commented:
@mm4tt feel free to lift the hold, just notice the PR got auto-approved after you lgtm'ed

ahg-g commented on clusterloader2/pkg/measurement/util/histogram.go:
```diff
@@ -28,6 +34,34 @@ type Histogram struct {
+		},
+	}
+
+	for k, v := range h.Buckets {
```
please add a comment clarifying that hist.Quantile expected a slice of Buckets ordered by UpperBound.

ingvagabund replied:
Done, right before calling sort.Slice

ahg-g commented:
/lgtm

mm4tt commented:
Something seems off,  `e2eSchedulingLatency` is being reported as a negative number:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1229767138370129920/artifacts/SchedulingMetrics_load_2020-02-18T14:23:50Z.json

Could you take a look?

ingvagabund commented:
> Could you take a look?

lemme see

ingvagabund commented:
@yiyang5055 the histogram built up from the samples is empty. That's why all the quantiles are `NaN` (resp. NaN converted into int64). Did you have a chance to actually test your changes?

ingvagabund commented:
Based on https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1230076412384251904/artifacts/SchedulingMetrics_load_2020-02-19T10:53:00Z.json there's supposed to be 16 samples. Wondering why there are not added to the histogram.

ingvagabund commented:
@yiyang5055 `e2eSchedulingMetricName` just needed to be prefixed with `scheduler_`. The histogram is populated correctly now.

ingvagabund commented:
All good now.

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1230141467285000192/artifacts/SchedulingMetrics_load_2020-02-19T15:17:05Z.json:
```
{
  "predicateEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "priorityEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "e2eSchedulingLatency": {
    "Perc50": 60473988,
    "Perc90": 335236626,
    "Perc99": 788254591
  }
}
```

ingvagabund commented:
/hold cancel

ahg-g commented:
/hold
why the other metrics are all zeros?

ingvagabund commented:
/test pull-perf-tests-clusterloader2

ingvagabund commented:
> why the other metrics are all zeros?
let's try once again

ahg-g commented:
> > why the other metrics are all zeros?
> 
> let's try once again

here is the PR history: https://prow.k8s.io/pr-history/?org=kubernetes&repo=perf-tests&pr=1061

The first three runs (from the right) had those metrics populated, the rest don't.

ingvagabund commented:
> The first three runs (from the right) had those metrics populated, the rest don't.

Will investigate. Though, I added only few new fields. Weird.

ingvagabund commented:
@mm4tt pull-perf-tests-clusterloader2 seems to be busted

ingvagabund commented:
Even here: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1064/pull-perf-tests-clusterloader2/1230826856261881856/artifacts/SchedulingMetrics_load_2020-02-21T12:34:08Z.json from https://github.com/kubernetes/perf-tests/pull/1064:
```json
{
  "predicateEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "priorityEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "e2eSchedulingLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  }
}
```

ahg-g commented:
right, you can see here as well: http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=predicate_evaluation

ahg-g commented:
/hold cancel
/lgtm

ahg-g commented:
/hold

we are probably now looking at deprecated metrics

mm4tt commented:
Interesting, it seems that it stopped working for master and 1.18 branch (which is daily rebased to master branch)
![cQAHfr3oK6P](https://user-images.githubusercontent.com/2604887/75094204-77fa6300-5589-11ea-8ec3-d90a8f9807fe.png)

Our older release tests are fine, e.g. 1.17:
![MR3W4CsfkiA](https://user-images.githubusercontent.com/2604887/75094235-9e200300-5589-11ea-8c52-dd1650d5e95c.png)

It looks like [these metrics are deprecated since 1.18](https://github.com/kubernetes/kubernetes/blob/d221d82eaf3756a759530bdd332af9ba755151dd/pkg/scheduler/metrics/metrics.go#L76), but they should be still populated, right? 

We can try debugging this, i.e. checking whether this is a scheduler issue (not-populating the metrics) or cl2 issue (not scraping them correctly), but given that they are deprecated, wouldn't be better to just switch to the successor metrics?

Anyway, I opened an issue for this - https://github.com/kubernetes/perf-tests/issues/1074

ingvagabund commented:
@ahg-g as from https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1076/pull-perf-tests-clusterloader2/1232004003802386432/artifacts/SchedulingMetrics_load_2020-02-24T18:34:12Z.json the metrics file looks like https://gist.github.com/ingvagabund/a9fc18c4ec0ddcfbf2aed84249aa447e.

```
$ cat FILE | grep scheduler_ | grep -v "# " | cut -d'{' -f1 | cut -d' ' -f1 | sort -u
scheduler_binding_duration_seconds_bucket
scheduler_binding_duration_seconds_count
scheduler_binding_duration_seconds_sum
scheduler_e2e_scheduling_duration_seconds_bucket
scheduler_e2e_scheduling_duration_seconds_count
scheduler_e2e_scheduling_duration_seconds_sum
scheduler_framework_extension_point_duration_seconds_bucket
scheduler_framework_extension_point_duration_seconds_count
scheduler_framework_extension_point_duration_seconds_sum
scheduler_pending_pods
scheduler_plugin_execution_duration_seconds_bucket
scheduler_plugin_execution_duration_seconds_count
scheduler_plugin_execution_duration_seconds_sum
scheduler_pod_preemption_victims_bucket
scheduler_pod_preemption_victims_count
scheduler_pod_preemption_victims_sum
scheduler_pod_scheduling_attempts_bucket
scheduler_pod_scheduling_attempts_count
scheduler_pod_scheduling_attempts_sum
scheduler_pod_scheduling_duration_seconds_bucket
scheduler_pod_scheduling_duration_seconds_count
scheduler_pod_scheduling_duration_seconds_sum
scheduler_queue_incoming_pods_total
scheduler_schedule_attempts_total
scheduler_scheduler_cache_size
scheduler_scheduler_goroutines
scheduler_scheduling_algorithm_duration_seconds_bucket
scheduler_scheduling_algorithm_duration_seconds_count
scheduler_scheduling_algorithm_duration_seconds_sum
scheduler_scheduling_algorithm_preemption_evaluation_seconds_bucket
scheduler_scheduling_algorithm_preemption_evaluation_seconds_count
scheduler_scheduling_algorithm_preemption_evaluation_seconds_sum
scheduler_total_preemption_attempts
scheduler_volume_scheduling_duration_seconds_bucket
scheduler_volume_scheduling_duration_seconds_count
scheduler_volume_scheduling_duration_seconds_sum
scheduler_volume_scheduling_stage_error_total
```

None of `scheduling_algorithm_predicate_evaluation_seconds` and `scheduling_algorithm_priority_evaluation_seconds` are reported. Only `binding_duration_seconds` is.

ingvagabund commented:
Due to https://github.com/kubernetes/kubernetes/pull/86584, both predicate/priority metrics are deprecated since 1.18:
- scheduling_algorithm_predicate_evaluation_seconds, replaced by framework_extension_point_duration_seconds[extenstion_point="Filter"]
- scheduling_algorithm_priority_evaluation_seconds, replaced by framework_extension_point_duration_seconds[extenstion_point="Score"]

mm4tt commented:
Thanks, @ingvagabund 

In that case, this change looks good to me.
/lgtm

ingvagabund commented:
/test pull-perf-tests-clusterloader2

ingvagabund commented:
Metrics are back!!! From https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/perf-tests/1061/pull-perf-tests-clusterloader2/1232251530732638208/artifacts/SchedulingMetrics_load_2020-02-25T10:56:11Z.json:
```
{
  "predicateEvaluationLatency": {
    "Perc50": 1303410,
    "Perc90": 6076030,
    "Perc99": 18711003
  },
  "priorityEvaluationLatency": {
    "Perc50": 3324628,
    "Perc90": 12249077,
    "Perc99": 31812986
  },
  "preemptionEvaluationLatency": {
    "Perc50": 0,
    "Perc90": 0,
    "Perc99": 0
  },
  "bindingLatency": {
    "Perc50": 32600018,
    "Perc90": 216257430,
    "Perc99": 463201886
  },
  "e2eSchedulingLatency": {
    "Perc50": 80327804,
    "Perc90": 406599501,
    "Perc99": 873739999
  }
}
```

ingvagabund commented:
@mm4tt metrics are popping back: http://perf-dash.k8s.io/#/?jobname=gce-100Nodes-master&metriccategoryname=Scheduler&metricname=SchedulingLatency&Operation=predicate_evaluation

ahg-g commented:
/hold cancel
